crates/admin/src/cluster_controller/logs_controller.rs

```diff
@@ -332,56 +333,210 @@ fn build_new_replicated_loglet_configuration(
+                .alive_nodes
+                .keys()
+                .copied()
+                .filter(|node_id| {
+                    nodes_config
+                        .get_log_server_storage_state(node_id)
+                        .can_write_to()
+                })
+                .collect(),
+        )
+    }
+
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+}
+
+/// Nodeset selector for picking a set of storage nodes for a replicated loglet out of a
+/// broader pool of available nodes.
+///
+/// This selector can be reused once constructed to make multiple decision in a single scheduling iteration,
+/// if the node configuration and the replication settings are not changing.
+#[cfg(feature = "replicated-loglet")]
+#[derive(Clone)]
+struct NodeSetSelector<'a> {
+    nodes_config: &'a NodesConfiguration,
+    cluster_state: &'a ObservedClusterState,
+}
+
+impl<'a> NodeSetSelector<'a> {
+    fn new<'b: 'a>(
+        nodes_config: &'b NodesConfiguration,
+        cluster_state: &'b ObservedClusterState,
+    ) -> NodeSetSelector<'a> {
+        Self {
+            nodes_config,
+            cluster_state,
+        }
+    }
+
+    /// Picks a set of storage nodes for a replicated loglet out of the available pool. Only alive, writable
+    /// storage nodes will be used.
+    fn select<R: Rng + ?Sized>(
+        &self,
+        strategy: NodeSetSelectionStrategy,
+        replication_property: &ReplicationProperty,
+        rng: &mut R,
+        preferred_nodes: &NodeSet,
+    ) -> Result<NodeSet, NodeSelectionError> {
+        if replication_property.at_greatest_scope().0 != &LocationScope::Node {
+            // todo: add support for other location scopes
+            unimplemented!("only node-scoped replication is currently supported");
+        }
+
+        // Only consider alive, writable nodes.
+        let candidates = WritableNodeSet::from(self.cluster_state, self.nodes_config);
+
+        if candidates.len() < replication_property.num_copies().into() {
+            return Err(NodeSelectionError::InsufficientWriteableNodes);
+        }
+
+        // ReplicationFactor(f+1) implies 2f+1 total nodes
+        assert!(
+            replication_property.num_copies() < u8::MAX >> 1,
+            "The implied nodeset size exceeds the maximum nodeset size"
+        );
+        let nodeset_target_size = (replication_property.num_copies() as usize - 1) * 2 + 1;
+        assert!(
+            nodeset_target_size >= replication_property.num_copies() as usize,
+            "The calculated target nodeset size can not be less than the replication factor"
+        );
+
+        let nodeset_target_size = match strategy {
+            NodeSetSelectionStrategy::InferredFaultToleranceStrict
+            | NodeSetSelectionStrategy::AllAvailable => nodeset_target_size,
+            NodeSetSelectionStrategy::CappedNodesetSize(size) => {
+                min(nodeset_target_size, size.get() as usize)
+            }
```

Comment:
What's the purpose of the `CappedNodesetSize` strategy? Going below the min nodeset size seems like a dangerous thing to do because one can immediately lose the ability to seal if one node goes down.

Reply:
My vision for it is that it would be configurable by an operator; this implies someone knows what they are doing and is opting into nodesets that never exceed a certain size. For example, I might want to run a large 20-node cluster but be comfortable with any one partition being replicated to 3 or 5 nodes, so I'd cap the size there. This is a relic of my first iteration from a couple of weeks back; let's take it out to avoid confusion.